magenta/interfaces/midi/midi.py

```diff
@@ -498,48 +516,216 @@ def stop_capture(self):
+        if msg.hex()[:5] == control_message.hex()[:5] and msg.type == 'note_on':
+          if msg.velocity != 0:
+            msg = msg.copy(velocity=1)
+        if msg.hex() == control_message.hex():
           return
     else:
-      # In a capture session.
-      control_tuple = ((control_number,) if control_value is None else
-                       (control_number, control_value))
-      if control_tuple not in self._control_cvs:
-        self._control_cvs[control_tuple] = threading.Condition(self._lock)
-      self._control_cvs[control_tuple].wait()
-
-  def start_playback(self, sequence, metronome_velocity):
+      # In a capture session
+      if control_message.hex() not in self._control_cvs:
+        self._control_cvs[control_message.hex()] = threading.Condition(
+          self._lock)
+      self._control_cvs[control_message.hex()].wait()
+
+  def start_playback(self, sequence):
     """Plays the monophonic, sorted NoteSequence through the MIDI output port.
 
     Stops any previously playing sequences.
 
     Args:
       sequence: The monohponic, chronologically sorted NoteSequence to play.
-      metronome_velocity: The velocity of the metronome's MIDI note_on message.
     """
     self.stop_playback()
-    self._player = MonoMidiPlayer(self._outport, sequence, metronome_velocity)
+    self._player = MonoMidiPlayer(self._outport, sequence)
     self._player.start()
+    self._inport.callback = self._capture_message
 
   def stop_playback(self):
     """Stops any active sequence playback."""
     if self._player is not None and self._player.is_alive():
       self._player.stop()
       stdout_write_and_flush('Stopped\n')
 
+  def execute_cc_message(self, message):
+    """Defines how to treat non Start/Stop user defined CC messages.
+
+    Args:
+      message: The incoming Control Change message.
+    """
+
+    # Metronome Velocity
+    if message.bytes()[:2] == _CC_MAP['Metronome velocity'].bytes()[:2]:
+      global _METRONOME_VELOCITY
+      _METRONOME_VELOCITY = message.bytes()[2]
+
+class CCRemapper(object):
+  """CC Message Remapping interface.
+
+  Attributes:
+    _inport: The Mido port for receiving messages.
+    _cc_map: The dictionary of user defined Control Change messages.
+
+  Args:
+    input_midi_port: The string MIDI port name to use for input.
+  """
+
+  def __init__(self, input_midi_port):
+    self._inport = mido.open_input(input_midi_port)
+    self._cc_map = _CC_MAP
+
+  def remapper_interface(self):
+    """Interface asking for user input of which parameters to remap"""
+
+    while True:
+      print "\nWhat would you like to assign a Control Change to...?:"
+      print "0. None (Exit)"
+      for i, CCparam in enumerate(sorted(self._cc_map.keys())):
+        print "{}. {}".format(i + 1, CCparam)
+      print ""
+      try:
+        cc_choice_index = int(raw_input("Please make a selection...\n>>>"))
+      except ValueError:
+        print "Please enter a number..."
+        time.sleep(1)
```

Comment:
because otherwise a lot of readable info goes by quickly. I took it out and if it bothersome without the pauses we can put it back in.

> On Sep 12, 2016, at 9:56 AM, Adam Roberts <notifications@github.com> wrote:
> 
> In magenta/interfaces/midi/midi.py <https://github.com/tensorflow/magenta/pull/209#discussion_r78410744>:
> 
> > +    self._cc_map = _CC_MAP
> > +
> > +  def remapper_interface(self):
> > +    """Interface asking for user input of which parameters to remap"""
> > +
> > +    while True:
> > +      print "\nWhat would you like to assign a Control Change to...?:"
> > +      print "0. None (Exit)"
> > +      for i, CCparam in enumerate(sorted(self._cc_map.keys())):
> > +        print "{}. {}".format(i + 1, CCparam)
> > +      print ""
> > +      try:
> > +        cc_choice_index = int(raw_input("Please make a selection...\n>>>"))
> > +      except ValueError:
> > +        print "Please enter a number..."
> > +        time.sleep(1)
> why sleep?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/tensorflow/magenta/pull/209/files/23b775fe695f914449d738c4e11817a65587fc8f#r78410744>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AP7D9QQtEWsd7qrrP_qsZMu77NlEVu0hks5qpYRXgaJpZM4J6Nk5>.
>